File: desktop/src/lib/etiquetasCodigosEan.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "../api/client";

export interface CodigoEan {
  id: string;
  sku: string;
  nombre_producto: string;
  numero_producto: number;
  presentacion: string;
  anio: number;
  bimestre: number;
  codigo: string;
  creado_at: string;
}

export interface NuevoCodigoEan {
  sku: string;
  nombre_producto?: string;
  numero_producto: number;
  presentacion?: string;
  anio?: number;
  mes: number;
}

export const BIMESTRE_LABEL: Record<number, string> = {
  0: "Ene-Feb",
  1: "Mar-Abr",
  2: "May-Jun",
  3: "Jul-Ago",
  4: "Sep-Oct",
  5: "Nov-Dic",
};

export function mesABimestre(mes: number): number {
  return Math.max(0, Math.min(5, Math.floor((mes - 1) / 2)));
}

/**
 * Menor número de producto libre en 1..900 (rellena huecos de EAN borrados).
 * Antes se usaba max+1 y los consecutivos liberados quedaban inaccesibles.
 */
export function siguienteNumeroProductoDisponible(
  codigos: Array<{ numero_producto?: number } | null | undefined>,
  minimo = 1,
  maximo = 900,
): number | null {
  const usados = new Set<number>();
  for (const c of codigos) {
    const n = Number(c?.numero_producto);
    if (Number.isFinite(n) && n >= minimo && n <= maximo) usados.add(n);
  }
  for (let n = minimo; n <= maximo; n++) {
    if (!usados.has(n)) return n;
  }
  return null;
}

/** Arma los 12 dígitos de datos (sin el verificador) para previsualizar el código. */
export function construirCodigo12(
  numeroProducto: number,
  presentacion: string,
  anio: number,
  bimestre: number,
): string {
  const prod = String(Math.max(1, Math.min(900, Math.round(numeroProducto) || 0))).padStart(3, "0");
  const pres = (presentacion || "000").replace(/\D/g, "").padStart(3, "0").slice(0, 3);
  const yy = String(((Math.round(anio) % 100) + 100) % 100).padStart(2, "0");
  const b = Math.max(0, Math.min(5, Math.round(bimestre) || 0));
  return `770${prod}${pres}${yy}${b}`;
}

/** Tamaños habituales; en SKUs compuestos (SHA70550mL) se toma el del final. */
const PRESENTACIONES_CONOCIDAS = [
  1000, 500, 400, 250, 150, 125, 120, 100, 60, 50, 40, 30, 20, 15, 10, 5,
] as const;

function codigoPresentacionDesdeNumero(n: number): string {
  if (n <= 0) return "000";
  if (n === 1000) return "001";
  if (n > 999) return String(n).slice(-3);
  return String(n).padStart(3, "0");
}

function numeroPresentacionDesdeCola(digitos: string): number | null {
  const raw = (digitos || "").replace(/\D/g, "");
  if (!raw) return null;
  if (raw.length <= 3) return Number(raw);
  for (const cand of PRESENTACIONES_CONOCIDAS) {
    const suf = String(cand);
    if (raw.endsWith(suf) && raw.length > suf.length) return cand;
  }
  return Number(raw.slice(-3));
}

/**
 * Infiere el código de presentación (3 dígitos) desde SKU/nombre.
 * kg / kilo / 1.000g / 1000g → 001; 50g → 050; 100g → 100; etc.
 */
export function sugerirPresentacionEan(sku: string, nombre = ""): string {
  const skuTrim = (sku || "").trim();
  const nombreTrim = (nombre || "").trim();
  const blob = `${skuTrim} ${nombreTrim}`.trim();

  if (/(?:1[.,]000|1000)\s*g\b/i.test(blob)) return "001";
  if (/kg\s*$/i.test(skuTrim) || /\b(?:kg|kilos?)\b/i.test(blob)) return "001";

  const colaSku = skuTrim.includes("-") ? skuTrim.slice(skuTrim.lastIndexOf("-") + 1) : skuTrim;
  if (
    /^(?:lt|l)$/i.test(colaSku) ||
    (/\b(?:lt|litro|litros)\b/i.test(blob) && !/\d+(?:[.,]\d+)?\s*(?:ml|lt|l)\b/i.test(blob))
  ) {
    return "001";
  }

  let m = skuTrim.match(/(\d+(?:[.,]\d+)?)\s*(g|ml|lt|l)\s*$/i);
  if (!m) m = blob.match(/(\d+(?:[.,]\d+)?)\s*(g|ml|lt|l)\b/i);
  if (!m) return "000";

  let unit = m[2].toLowerCase();
  if (unit === "l") unit = "lt";
  const digitos = String(m[1]).replace(",", ".").split(".")[0];
  if (unit === "kg") return "001";
  if (unit === "lt" && digitos === "1") return "001";

  const n = numeroPresentacionDesdeCola(digitos);
  if (n == null) return "000";
  if (unit === "g" && n === 1000) return "001";
  return codigoPresentacionDesdeNumero(n);
}

const QUERY_KEY = ["etiquetas-codigos-ean"];

export function useCodigosEan() {
  return useQuery({
    queryKey: QUERY_KEY,
    queryFn: async () => {
      const data = await api.get<{ codigos: CodigoEan[] }>("/api/etiquetas/codigos-ean");
      return data.codigos ?? [];
    },
    staleTime: 30_000,
  });
}

export function useCrearCodigoEan() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (nuevo: NuevoCodigoEan) =>
      api.post<CodigoEan & { ok: boolean }>("/api/etiquetas/codigos-ean", nuevo),
    onSuccess: () => void qc.invalidateQueries({ queryKey: QUERY_KEY }),
  });
}

export function useActualizarCodigoEan() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, datos }: { id: string; datos: NuevoCodigoEan }) =>
      api.put<CodigoEan & { ok: boolean }>(`/api/etiquetas/codigos-ean/${id}`, datos),
    onSuccess: () => void qc.invalidateQueries({ queryKey: QUERY_KEY }),
  });
}

export function useEliminarCodigoEan() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => api.delete<{ ok: boolean }>(`/api/etiquetas/codigos-ean/${id}`),
    onSuccess: () => void qc.invalidateQueries({ queryKey: QUERY_KEY }),
  });
}

export interface ResultadoImportacionEan {
  ok: boolean;
  creados: number;
  omitidos: number;
  errores: string[];
  siguiente_numero?: number;
}

/** Registra en bloque los combos SIIGO activos que aún no tienen EAN. */
export function useImportarCombosEanSiigo() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: () =>
      api.post<ResultadoImportacionEan>("/api/etiquetas/codigos-ean/importar-siigo", {}),
    onSuccess: () => void qc.invalidateQueries({ queryKey: QUERY_KEY }),
  });
}

export interface ResultadoSyncBarcodeSiigo {
  ok: boolean;
  actualizados: number;
  omitidos: number;
  errores: string[];
  procesados?: number;
  en_planilla?: number;
}

/** Empuja los EAN de la planilla al campo código de barras en SIIGO. */
export function useSincronizarBarcodesEanSiigo() {
  return useMutation({
    mutationFn: (opts?: { solo_vacios?: boolean }) =>
      api.post<ResultadoSyncBarcodeSiigo>("/api/etiquetas/codigos-ean/sincronizar-siigo", {
        solo_vacios: opts?.solo_vacios ?? true,
      }),
  });
}

File: app/tools/etiquetas_codigos_ean.py
"""Helpers para códigos EAN-13 internos (planilla de etiquetas).

Estructura: 770 + producto(3) + presentación(3) + año(2) + bimestre(1) + check.
La presentación codifica el contenido neto:
  - kg / kilo / 1 kg / 1.000 g / 1000 g → 001
  - 50 g / 50 mL → 050
  - 5 mL → 005
  - 100 g → 100
  - valores de 3 dígitos se usan tal cual (125, 250, 500…)
  - SKUs compuestos (ej. C-SHA70550mL) toman el tamaño real del final (50, no 70550)
"""
from __future__ import annotations

import re
from datetime import datetime
from typing import Iterable

# Tamaños comunes; se prueban de mayor a menor longitud al final del número del SKU.
_PRESENTACIONES_CONOCIDAS = (
    1000, 500, 400, 250, 150, 125, 120, 100, 60, 50, 40, 30, 20, 15, 10, 5,
)


def normalizar_sku_ean(sku: str) -> str:
    return re.sub(r"\s+", "", (sku or "").strip().upper())


def _codigo_presentacion_desde_numero(n: int) -> str:
    if n <= 0:
        return "000"
    if n == 1000:
        return "001"
    if n > 999:
        return str(n)[-3:]
    return f"{n:03d}"


def _numero_presentacion_desde_cola(digitos: str) -> int | None:
    """Si el número pega marca+tamaño (70550), toma el tamaño final conocido."""
    raw = re.sub(r"\D", "", digitos or "")
    if not raw:
        return None
    if len(raw) <= 3:
        return int(raw)
    for cand in _PRESENTACIONES_CONOCIDAS:
        suf = str(cand)
        if raw.endswith(suf) and len(raw) > len(suf):
            return cand
    return int(raw[-3:])


def presentacion_ean_desde_sku(sku: str, nombre: str = "") -> str:
    """Infiere el código de 3 dígitos de presentación a partir del SKU/nombre."""
    sku = (sku or "").strip()
    nombre = (nombre or "").strip()
    blob = f"{sku} {nombre}".strip()

    # 1.000 g / 1000 g → 001 (equivalente a 1 kg)
    if re.search(r"(?:1[.,]000|1000)\s*g\b", blob, re.I):
        return "001"

    # Kg / kilo al final del SKU o en el texto (cualquier presentación en kilo → 001)
    if re.search(r"kg\s*$", sku, re.I) or re.search(r"\b(?:kg|kilos?)\b", blob, re.I):
        return "001"

    # Litro suelto (Lt / L sin cantidad) → 001 (unidad completa)
    cola = sku.split("-")[-1] if "-" in sku else sku
    if re.search(r"(?:lt|l)\s*$", sku, re.I) and not re.search(r"\d", cola):
        return "001"
    if re.search(r"\b(?:lt|litro|litros)\b", blob, re.I) and not re.search(
        r"\d+(?:[.,]\d+)?\s*(?:ml|lt|l)\b", blob, re.I
    ):
        return "001"

    # Preferir cantidad al final del SKU (convención SIIGO: …250g, …50mL)
    m = re.search(r"(\d+(?:[.,]\d+)?)\s*(g|ml|lt|l)\s*$", sku, re.I)
    if not m:
        m = re.search(r"(\d+(?:[.,]\d+)?)\s*(g|ml|lt|l)\b", blob, re.I)
    if not m:
        return "000"

    unit = m.group(2).lower()
    if unit == "l":
        unit = "lt"
    digitos = m.group(1).replace(",", ".").split(".")[0]

    if unit == "kg":
        return "001"
    if unit == "lt" and digitos in {"1", "1.0", "1.00"}:
        return "001"

    n = _numero_presentacion_desde_cola(digitos)
    if n is None:
        return "000"
    if unit == "g" and n == 1000:
        return "001"
    return _codigo_presentacion_desde_numero(n)


def siguiente_numero_producto(numeros: Iterable[int], minimo: int = 1, maximo: int = 900) -> int | None:
    """Menor número libre en [minimo, maximo] (rellena huecos de códigos borrados)."""
    usados: set[int] = set()
    for n in numeros:
        try:
            v = int(n)
        except (TypeError, ValueError):
            continue
        if minimo <= v <= maximo:
            usados.add(v)
    for n in range(minimo, maximo + 1):
        if n not in usados:
            return n
    return None


def anio_bimestre_actual(ahora: datetime | None = None) -> tuple[int, int]:
    ahora = ahora or datetime.now()
    anio = ahora.year % 100
    bimestre = max(0, min(5, (ahora.month - 1) // 2))
    return anio, bimestre
